File: Data Structures & Algorithms/house-robber/submission-2.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        vector<int> dp(nums.size(),-1);
        return helper(nums.size(),0, nums,dp);
    }

    int helper(int n, int i, vector<int>& nums, vector<int> &dp) {
        if(i>=n) return 0;

        if(dp[i]!=-1) return dp[i];
        return dp[i] =  max(helper(n, i+1, nums,dp), nums[i]+helper(n, i+2, nums,dp));
    }
};

File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-2.java
public class Solution {
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length <= 1) return 0;

        int held = -prices[0]; // dp[i-1][1]
        int notHeld = 0;       // dp[i-1][0]
        int notHeldPrev = 0;   // dp[i-2][0]

        for (int i = 1; i < prices.length; i++) {
            int nextNotHeld = Math.max(notHeld, held + prices[i]);
            int nextHeld = Math.max(held, notHeldPrev - prices[i]);
            
            // Shift variables forward for the next day's iteration
            notHeldPrev = notHeld;
            notHeld = nextNotHeld;
            held = nextHeld;
        }

        return notHeld;
    }
}